### ISSUE TITLE
KS using .predict_proba() and .score() of Scorecard is different

### ISSUE THREAD

**Issue by ds-mauri** [guillermo-navas-palencia/optbinning#184]
Hi @guillermo-navas-palencia !

As always, congratulations for the amazing job on the library! 
I was doing some tests using the Scorecard class and noticed that de KS statistics reported when using the method .predict_proba and .score are differents. I was wondering if that was the expected behaviour. The score transformation wasn't supposed to be a linear transformation of the weights of the logistic model? Thought this shouldn't modify the KS statistics, but I'm not sure.

Best regards,

![image](https://user-images.githubusercontent.com/34061728/184031564-abf87a84-c456-49a0-abe1-b411bb9282a8.png)

**Comment by ds-mauri** [comment]
UPDATE

Indeed, even when computing the AUROC the metric changes if we use the .predict_proba() or .score(). To further investigate that I ploted the bads using the deciles of each "score" along the x-axis. For some reason the bad rate changes for each bin. On the left I used the .predict_proba() method (rescaled to 0-1000) and on the right the .score().

![image](https://user-images.githubusercontent.com/34061728/184227622-7b44df98-81fa-44b1-bd8f-6272dee8a46c.png)

Checking on a customer level we see that for some users the score transformation change their ordenation on the dataset.

![image](https://user-images.githubusercontent.com/34061728/184230768-66fc47a9-b376-4b06-88ba-0e34c42f300a.png)

Here I take 3 customers as example. Using the .predict_proba() they were ranked as 6th and 3rd decile, respectively. Using the .score(), both was on the 4th decile. Is this the expected behaviour of the Scorecard transformation or it should keep the ordering? I believe this is more a question of the theory behind than an issue of the implementation.

Thanks again!

**Comment by guillermo-navas-palencia** [comment]
Hi @ds-mauri!

Interesting, I will investigate during the weekend. Could you please specify the version you are using?

**Comment by ds-mauri** [comment]
Sure! I'm using 0.14.1. Just saw on release notes that was a bugfix on version 0.15.0 on the .score() method. I'll check if it's the case.

**Comment by guillermo-navas-palencia** [comment]
Hi @ds-mauri,

The relationship between score and probability is non-linear: ``np.log((1. / event_rate - 1) * n_event / n_nonevent)``. See the following example:

```python
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from optbinning import BinningProcess
from optbinning import Scorecard
from optbinning.scorecard import plot_auc_roc, plot_cap, plot_ks

from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression

data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

binning_process = BinningProcess(data.feature_names)

estimator = LogisticRegression(solver="lbfgs", class_weight="balanced")

scorecard = Scorecard(binning_process=binning_process,
                      estimator=estimator, scaling_method="min_max",
                      scaling_method_params={"min": 300, "max": 850})

scorecard.fit(X, y)
```

Plots:
```python
y_pred = scorecard.predict_proba(X)[:, 1]
y_score = scorecard.score(X)

plot_ks(y, y_pred)
plot_ks(y, y_score)
```
![image](https://user-images.githubusercontent.com/40094942/186385974-cf259d96-103c-4b1e-a2aa-c15229285533.png)
![image](https://user-images.githubusercontent.com/40094942/186386014-afa84eb6-7ce2-4048-a357-59ab95b7157a.png)


Relationship between score and predicted probability
```python
plt.scatter(y_pred, y_score)
plt.show()
```
![image](https://user-images.githubusercontent.com/40094942/186386085-3f1a3e68-0fdf-4471-a6fd-a4f33a16ac6a.png)


Non-linear transformation
```python
n_event = y.sum()
n_nonevent = len(y) - n_event

score = np.log((1. / y_pred - 1) * n_event / n_nonevent)

plt.scatter(score, y_score)
plt.show()

plot_ks(y*-1 + 1, score)
```
![image](https://user-images.githubusercontent.com/40094942/186386213-d9c48be1-c499-4646-8a20-3d9c329117ed.png)
![image](https://user-images.githubusercontent.com/40094942/186386243-bc6e2608-6247-4f84-b511-7582e375c44b.png)

**Comment by guillermo-navas-palencia** [comment]
The transformation from the score to PD and vice-versa can be found here: https://github.com/guillermo-navas-palencia/optbinning/blob/master/optbinning/binning/transformations.py

**Comment by ds-mauri** [comment]
Hi @guillermo-navas-palencia ! 

Thanks a lot for the amazing explanation. Indeed, I made a little bit of confusion using both methods. I said "linear" but I was intending to say "monotonic" instead. In fact, it isn't a linear transformation at all, thanks!

Regarding the concern that the .score method was changing the ranking of the scored instances in comparison with the .predict_proba, it was my mistake in validating the bugfix that you uploaded in version 0.15.0. So, in version 0.15.0 the .score method is perfectly implemented.

Just to make it clear:
- as the .score method isn't a linear transformation of the PD, it changes the underlying distribution and this changes the KS (the AUROC, on the other hand, keeps the same)
- the ranking of both methods is the same. So, the instance with the worst score also should have the worst PD and so on.

By the way, I didn't complete catch the score transformation `np.log((1. / event_rate - 1) * n_event / n_nonevent)`. This is somehow equivalent to the "classical" method without the scaling parameters?
![image](https://user-images.githubusercontent.com/34061728/186482720-81011597-5842-488c-a003-4df5c2e0c2c6.png)

Best regards,

**Comment by guillermo-navas-palencia** [comment]
Hi @ds-mauri,

Regarding the score transformation, I wrote the logarithmic transformation to convert PD to WoE as a similar example. Therefore, the score is your expression above replacing woe by np.log((1. / PD - 1) * n_event / n_nonevent).

**Comment by ds-mauri** [comment]
Oh, it's clear! Thanks again @guillermo-navas-palencia !
